Add storage pool configuration as yaml

edlerd opened the pull request:
## Done

- Add storage pool configuration as yaml
- Hide cursor on yaml forms in read only mode
- Add prompt on switching between yaml and guided forms for networks, profiles, instances and storage pools if the yaml was edited
- Add notification with link to relevant docs page from the yaml config
- refresh yaml editor, when having changes in it and "cancel"
- fix profile edit, when doing changes and going to yaml, the yaml now represents the changed values from the form

Fixes WD-8614

## QA

1. Run the LXD-UI:
    - On the demo server via the link posted by @webteam-app below. This is only available for PRs created by collaborators of the repo. Ask @mas-who or @edlerd for access.
    - With a local copy of this branch, run as described [here](../HACKING.md#setting-up-for-development).
2. Perform the following QA steps:
    - create a storage pool, switch to YAML configuration, ensure it works as intended
    - edit  a storage pool, switch to YAML configuration, ensure it works as intended

webteam-app commented:
Demo starting at https://lxd-ui-676.demos.haus

mas-who commented:
Testing on the demo server results in the below behaviour:

![Screenshot from 2024-02-27 09-07-10](https://github.com/canonical/lxd-ui/assets/151511568/e5f2fbfa-9a9c-4c3d-9b0d-e863ca73565b)

The above error actually occurs when creating a pool normally so it's not limited to the yaml editor. However, the pool actually gets created as shown below. Navigating to the pool configuration page works and updating the pool via the yaml editor also works perfectly.

![Screenshot from 2024-02-27 09-08-00](https://github.com/canonical/lxd-ui/assets/151511568/73e76ecd-e381-48c0-9efd-0eb215ee8531)

I then tested locally and everything works, so I think something might be off on the demo server, maybe it's got to do with the OS update from last week?

edlerd commented:
> I then tested locally and everything works, so I think something might be off on the demo server, maybe it's got to do with the OS update from last week?

This error happens when the demo server is slow. We have a timeout in haproxy, and if that hits, you get the above error as the api request will be a haproxy error page that is not valid json.

mas-who commented:
> > I then tested locally and everything works, so I think something might be off on the demo server, maybe it's got to do with the OS update from last week?
> 
> This error happens when the demo server is slow. We have a timeout in haproxy, and if that hits, you get the above error as the api request will be a haproxy error page that is not valid json.

That's good to know, thanks! Then I think it's LGTM :)

piperdeck commented:
I have a few thoughts about this, which I think could apply to YAML configuration in the UI generally.

# 1. 

The first is that there's a bit of a frustrating interaction where you click on the YAML, expecting to make some changes, and you get the error saying "cannot edit in read-only editor". I think we should make it more clear that the editor is read-only before the user has a chance to try to type in it.

- Desaturate the whole YAML box while it's in read-only mode, only setting it to full saturation when it's in edit mode
- Disable the text cursor from appearing when the editor is in read-only mode. If I click on a box and I see a text cursor appear, I expect to be able to input text at the location of the cursor
- Of course, the text inside should always be selectable, but perhaps the selection highlighting should be browser/OS native when the box is in read-only mode

# 2.

The inline message that appears above the text box is a little awkward. 

- Is it saying that if I ever go back to the inline forms and change anything, it will reset all the changes I made to the YAML? That's not consistent with a brief test of increasing the size through the YAML and then going back to the inline forms. When I do that, the change I made in the YAML seems to persist. 
- Is it saying that the YAML is always regenerated from the entity's existing configuration every time you open it? Then I would say that instead. I'd like to propose a new wording for this warning but I'm not sure exactly what the technical details are so I'm not sure.
- Is it saying that the YAML config will be lost, but only if I switch back to guided forms before saving? In this case, we could simply show the user a confirmation modal in the event that they click on the guided forms while they have unsaved changes in the YAML

# 3.

A user who switches to YAML configuration is choosing to lose some convenience, but maybe we can make it easier by showing them which YAML configurations are available through commented out config lines? example:

```
name: abc
description: ''
driver: zfs
source: /var/snap/lxd/common/lxd/disks/abc.img
config:
  size: 31GiB

# Uncomment to allow foo operations rather than regular operations
# foo: true

# Uncomment to choose the number of bars in pool
# bar: 0
```

edlerd commented:
> * Desaturate the whole YAML box while it's in read-only mode, only setting it to full saturation when it's in edit mode

I am not sure how to do this. I played locally with reduced opacity, but this doesn't transport the "this can't be edited" message. We could use a different theme for the editor, but that also seems a bit odd and doesn't transport the wanted message in my opinion.

> * Disable the text cursor from appearing when the editor is in read-only mode. If I click on a box and I see a text cursor appear, I expect to be able to input text at the location of the cursor

This should now work as proposed.

> The inline message that appears above the text box is a little awkward.

We can work around this, I changed how networks and storage pools work, so we translate the yaml form back to the guided forms. Also removed the note as it just works now to switch back and forth.

> A user who switches to YAML configuration is choosing to lose some convenience, but maybe we can make it easier by showing them which YAML configurations are available through commented out config lines? example:

The idea of the YAML forms is to allow adding options that we don't know about in the UI. It seems counterintuitive to maintain a list of possible keys and descriptions. We could use the metadata api, but it seems rather complex to build. This should be a dedicated ticket tough, as it is rather involved to build. A cheap alternative would be to link to the relevant page in the docs. Which might be the better thing to do anyway, as the docs can provide more info on the keys.

piperdeck commented:
> I am not sure how to do this. I played locally with reduced opacity, but this doesn't transport the "this can't be edited" message. We could use a different theme for the editor, but that also seems a bit odd and doesn't transport the wanted message in my opinion

Playing around with it, I think that removing the text cursor will suffice

> We can work around this, I changed how networks and storage pools work, so we translate the yaml form back to the guided forms. Also removed the note as it just works now to switch back and forth.

Wow, this works really well. I didn't know this solution was possible haha

However, it looks like this introduces a new issue, because if I change the name or driver of a pool in the YAML editor, those changes are adopted by the disabled fields in the guided form, and the "save changes" button is disabled. Maybe instead, if you're switching from the YAML editor to the forms, any changes that would alter the frozen fields should be thrown out.

Otherwise, it might just be safer to have a warning appear when you try to switch back to the forms, saying 

"Switching back to guided forms will discard all changes in the YAML editor. Are you sure you want to proceed?

Cancel
Proceed and discard changes"

> The idea of the YAML forms is to allow adding options that we don't know about in the UI. It seems counterintuitive to maintain a list of possible keys and descriptions. We could use the metadata api, but it seems rather complex to build. This should be a dedicated ticket tough, as it is rather involved to build. A cheap alternative would be to link to the relevant page in the docs. Which might be the better thing to do anyway, as the docs can provide more info on the keys.

A fair point. I think linking to the docs would probably be a suitable solution here.

edlerd commented:
> However, it looks like this introduces a new issue, because if I change the name or driver of a pool in the YAML editor, those changes are adopted by the disabled fields in the guided form, and the "save changes" button is disabled. Maybe instead, if you're switching from the YAML editor to the forms, any changes that would alter the frozen fields should be thrown out.
> 
> Otherwise, it might just be safer to have a warning appear when you try to switch back to the forms, saying
> 
> "Switching back to guided forms will discard all changes in the YAML editor. Are you sure you want to proceed?
> 
> Cancel Proceed and discard changes"

Yes, a prompt is better because of various problems the user could cause when editing yaml. I changed it for networks, profiles, instances and storage pools

Also fixed a bunch of other problems with yaml configuration (see description) and added the link to the docs from the yaml config pages.

edlerd commented:
Implemented the discussed changes, @piperdeck can you have another look at this?
